gostream/codec/h264/encoder.go

```diff
@@ -0,0 +1,148 @@
+)
+
+type encoder struct {
+	img     image.Image
+	reader  video.Reader
+	codec   *avcodec.Codec
+	context *avcodec.Context
+	width   int
+	height  int
+	frame   *avutil.Frame
+	pts     int64
+	logger  golog.Logger
+}
+
+func (h *encoder) Read() (img image.Image, release func(), err error) {
+	return h.img, nil, nil
+}
+
+// NewEncoder returns an h264 encoder that can encode images of the given width and height. It will
+// also ensure that it produces key frames at the given interval.
+func NewEncoder(width, height, keyFrameInterval int, logger golog.Logger) (codec.VideoEncoder, error) {
+	h := &encoder{width: width, height: height, logger: logger}
+
+	if h.codec = avcodec.FindEncoderByName(V4l2m2m); h.codec == nil {
+		return nil, errors.Errorf("cannot find encoder '%s'", V4l2m2m)
+	}
+
+	if h.context = h.codec.AllocContext3(); h.context == nil {
+		return nil, errors.New("cannot allocate video codec context")
+	}
+
+	h.context.SetEncodeParams(width, height, avcodec.PixelFormat(pixelFormat), false, 10)
+	h.context.SetTimebase(1, keyFrameInterval)
+
+	h.reader = video.ToI420((video.ReaderFunc)(h.Read))
+
+	if h.context.Open2(h.codec, nil) < 0 {
+		return nil, errors.New("cannot open codec")
+	}
+
+	if h.frame = avutil.FrameAlloc(); h.frame == nil {
+		h.context.Close()
+		return nil, errors.New("cannot alloc frame")
+	}
+
+	return h, nil
+}
+
+func (h *encoder) Encode(ctx context.Context, img image.Image) ([]byte, error) {
+	if err := avutil.SetFrame(h.frame, h.width, h.height, pixelFormat); err != nil {
+		return nil, errors.Wrap(err, "cannot set frame properties")
+	}
+
+	if ret := avutil.FrameMakeWritable(h.frame); ret < 0 {
+		return nil, errors.Wrap(avutil.ErrorFromCode(ret), "cannot make frame writable")
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	h.img = img
+	yuvImg, release, err := h.reader.Read()
+	defer release()
+
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot read image")
+	}
+
+	h.frame.SetFrameFromImg(yuvImg.(*image.YCbCr))
+	h.frame.SetFramePTS(h.pts)
+	h.pts++
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+		return h.encodeBytes(ctx)
+	}
+}
+
+func (h *encoder) encodeBytes(ctx context.Context) ([]byte, error) {
+	pkt := avcodec.PacketAlloc()
+	if pkt == nil {
+		return nil, errors.New("cannot allocate packet")
+	}
+	defer pkt.Free()
+	defer pkt.Unref()
+	defer avutil.FrameUnref(h.frame)
+
+	if ret := h.context.SendFrame((*avcodec.Frame)(unsafe.Pointer(h.frame))); ret < 0 {
+		return nil, errors.Wrap(avutil.ErrorFromCode(ret), "cannot supply raw video to encoder")
+	}
+
+	var bytes []byte
+	var ret int
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		ret = h.context.ReceivePacket(pkt)
+		switch ret {
+		case avutil.Success:
+			payload := C.GoBytes(unsafe.Pointer(pkt.Data()), C.int(pkt.Size()))
+			bytes = append(bytes, payload...)
+			pkt.Unref()
+		case avutil.ErrorEAGAIN:
+			break loop
```

Comment:
One question here on error handling. 

Given that this loop appears to accumulate the packets for a given encoded frame. Is it possible that a one-off `EAGAIN` occurs from `ReceivePacket` but the next one is successful? In such a case, I am wondering if it makes more sense to continue here.

Reply:
Good question. Here, I'm just doing exactly what the docs are telling me to do. From [here](https://ffmpeg.org/doxygen/3.4/group__lavc__encdec.html)
> For encoding, call [avcodec_receive_packet()](https://ffmpeg.org/doxygen/3.4/group__lavc__decoding.html#ga5b8eff59cf259747cf0b31563e38ded6). On success, it will return an [AVPacket](https://ffmpeg.org/doxygen/3.4/structAVPacket.html) with a compressed frame. Repeat this call until it returns [AVERROR(EAGAIN)](https://ffmpeg.org/doxygen/3.4/group__lavu__error.html#gae4bb6f165973d09584e0ec0f335f69ca) or an error. The [AVERROR(EAGAIN)](https://ffmpeg.org/doxygen/3.4/group__lavu__error.html#gae4bb6f165973d09584e0ec0f335f69ca) return value means that new input data is required to return new output.

If something goes wrong here it shouldn't throw an EAGAIN error but another error. That other error we'll return.

Reply:
I added that link as a comment to this loop.